acceptancetests/assess_persistent_storage.py

```diff
@@ -27,6 +33,116 @@
+    """This function tests charm deployment and removal with single
+       persistent storage unit.
+
+       Steps taken to the test:
+       - Deploy dummy-storage charm with a single block storage unit (ebs)
+         and a single filesystem storage unit (rootfs).
+       - Check charm status once the deployment is done.
+       - Check charm storage list once the deployment is done.
+       - Remove the deployed charm, the output should indicate that the
+         rootfs storage unit will be removed, ebs storage unit will be
+         detached.
+       - Check storage list again to ensure ebs storage units really exists.
+
+       :param client: ModelClient object to deploy the charm on.
+    """
+
+    charm_name = 'dummy-storage'
+    charm_path = local_charm_path(
+        charm=charm_name, juju_ver=client.version)
+    log.info(
+        '{} is going to be deployed with 1 ebs block storage unit and '\
+        '1 rootfs filesystem storage unit.'.format(charm_name))
+    log.info(
+        'juju deploy ${JUJU_REPOSITORY}/charms/dummy-storage '\
+        '--storage single-blk=ebs --storage single-fs=rootfs')
+    # client.deploy() doesn't support repeat parameters,
+    # although juju deploy supports it
+    #client.deploy(
+    #    charm_path,
+    #    storage='single-blk=ebs',
+    #    storage='single-fs=rootfs')
+    client.get_juju_output(
+        'deploy', '/home/noname/work/Canonical/workbench/repository/charms/dummy-storage',
+        '--storage', 'single-blk=ebs',
+        '--storage', 'single-fs=rootfs',
+        include_e=False)
+    client.wait_for_started()
+    client.wait_for_workloads()
+
+    log.info('juju status --format json')
+    status_output = json.loads(
+        client.get_juju_output('status', '--format', 'json', include_e=False))
+    app_status = status_output['applications'][charm_name]['application-status']['current']
+
+    if app_status != 'active':
+        log.error(
+            'Charm {} is not functioning properly, test terminated.'.format(
+            charm_name))
+        sys.exit(1)
+    else:
+        log.info('juju storage --format json')
+        storage_output = json.loads(
+            client.get_juju_output('storage', '--format', 'json', include_e=False))
+
+        # check the total number of storage unit(s) and name(s)
+        storage_list = storage_output['storage'].keys()
+        if len(storage_list) != 2:
+            log.error(
+                '{} storage unit(s) found, should be 2, test terminated.'.format(
+                str(len(storage_list))))
+            sys.exit(1)
+        else:
+            log.info(
+                'Following storage units have been found:\n{}'.format(
+                '\n'.join(storage_list)))
+            single_fs_id = ''
+            single_blk_id = ''
+            for elem in storage_list:
+                if elem.startswith('single-fs'):
+                    single_fs_id = elem
+                    log.info(
```

Comment:
For the storage number and name, there is a story. In charm storage test, an unexpected additional storage item shows up, say, expected 2 units but got 3. Lesson learned from there to make me add the total number check to detect the problem in a easiest way. For the storage name check, in system, there are storage provider id, storage name/id and storage index - /0, /1, etc. This check is to ensure although the storage index can change, the prefix of name should be the same - it would be a bug if somehow the storage name/id and provider id swap around. For the log level, considering the importance of this part, I'd like to set it to info() to show up in normal test run.